### ISSUE TITLE
Main DNS FQDN is invalid

### ISSUE THREAD

**Issue by hunter-nl** [Ylianst/MeshCentral#366]
With recent update(s), the (new deployed) agents pointing to example.com instead of mc.example.com. The server itself is now running also on example.com instead of the FQDN. I'm running in TLSoffload reverse proxy mode, so Cert config item is not possible to configure.
With default domain section, a DNS entry is not allowed.
A sub entry of domain, a DNS entry is possible and is working fine.

How to solve this or manually configure the correct FQDN to be used as seen 'outside' the reverse proxy?

Or is the processing of the certificate changed? I use a wildcard certificate and this will be put into the data folder when certificate gets renewed. Th CN is example.com but the SAN contains *.example.com and example.com (as specified by letsencrypt). So, how can I now specify mc.example.com to be used by the agents and server (server use it for example the invitation link)?

**Comment by Ylianst** [comment]
Hi. First, make sure to backup your server before doing this change. When, just add the "cert" value in the settings section of config.json with the external name of your server like this:

```
{
  "settings": {
    "Cert": "devbox.mesh.meshcentral.com",
    "Port": 443
  }
}
```

Some of the self-signed certificates will be re-generated on the next server run, after that you should be fine. Let me know if that works. Your right that for the default domain, the "cert" name is used, you can't set a "dns" entry in the default domain.

Hope it helps,
Ylian

**Comment by hunter-nl** [comment]
It should not create the self signed cert, but use the provided cert (as used on the HAProxy). As the provided cert will be seen by the agents. Cert_url item in config does not work with HAproxy (Chicken and egg order issue). Without the correct cert, I get hash agent errors. It would be nice if default domain can have a DNS item.

**Comment by Ylianst** [comment]
So, you still need to set "cert":"name" to the name your server will be using, even if you don't plan on using that self-signed cert. After that, the server will use that DNS name as the default name to tell agents where to connect to. Also, as a temporary measure, you can set

`"ignoreagenthashcheck": true`

In the settings section to have agents connect even if the cert hash is not right.

**Comment by hunter-nl** [comment]
But ignoring agent hash is an additional security risk. Can generating self-signed certs set disabled and use "cert":"fqdn" option (or similar option to set the FQDN to be used as default)?

**Comment by hunter-nl** [comment]
I tried to use the "cert":"name" entry. When not self signed cert is already in place. It still use the common name of that cert file. Not the name from config file as specified in "cert":"name". And the cert common name is not the name of the FQDN to reach the server (through reverse proxy). In SAN attribute of the cert the wildcard is specified (which is the correct way according to letsencrypt). So from the cert it is impossible to determine the FQDN to be used as default MeshCentral server name. Can de program code be adjusted to allow a config setting to set the correct name te be used for the default domain? Of course the specified name must "fit" into the wildcard name. So MeshCentral.example.com as name must match *.example.com.

**Comment by hunter-nl** [comment]
This is still not fixed. Please remove the fixed label. MeshCentral choose wrong FQDN when wildcard certificate is in use. The name in common name is not always the correct DNS name to be used. So please make it configurable for the default domain (the other domains can have a DNS configuration item). Agents getting now the wrong url in the configuration file. And links for example to start messaging or the invitation link are now wrong.

**Comment by Ylianst** [comment]
Arg. Ok, I will need to investigate this. I also want to be able to deal with a reverse-proxy changing the TLS cert on-the-fly, without MeshCentral server restart. So, I will probably look at all this at this same time.

**Comment by JamesLodge** [comment]
I'm also using a wildcard cert (letsencrypt) on my Nginx r-proxy and seeing Agent bad web cert hash. I assume my issue is the same as above?

**Comment by JamesLodge** [comment]
What is the solution for this issue?

**Comment by hunter-nl** [comment]
The only solution I see is when @Ylianst is implementing a configuration option to set the primary FQDN to be used instead retrieve the CN=*.domain.tld from certificate.

**Comment by JamesLodge** [comment]
@Ylianst is this possible? can you allow the configuration of FQDN rather than pulling from the cert? I've messaged you a few times on Twitter regarding this. Its common place (i'd say) that having Nginx reverse proxy using a wildcard cert as you'd proxy many services, not just Mesh Central.

**Comment by Ylianst** [comment]
I just published MeshCentral v0.4.3-n with one fix related to this. When MeshCentral loads the certificate from the reverse proxy using the "CertUrl" setting, it was not setting the hostname in the request. If the reverse proxy was using many different certificates, it would not know which one to return.

In the middle of the following screen, MeshCentral loads the certificates for both my domains and indicates that hostname it's using to make the request.

![x](https://user-images.githubusercontent.com/1319013/67796861-fdd42a80-fa3d-11e9-9fd5-161f27778b14.png)

**Comment by hunter-nl** [comment]
You do not understand the issue correctly. :-)
Let me try to explain.

The reverse proxy (in my case HAproxy), use SNI. When MC server try to reach the reserve proxy with IP address the reverse proxy can not match any available certificates (including wildcards) and returns a service not available HTML error code back.
Secondly: MC server is doing this check before http service of MC server is up and running. The reverse proxy checks frequently the ‘health’ of MC server. When http of MC server is still not running then reverse proxy will not accept any request for the MC server. So even if MC server is using a FQDN request to get the CN from reverse proxy certificate(s). It still do not get it from reserve proxy as it returns service unavailable HTML error code.
Third: MC server use the CN from the provided certificate. And the CN is the wildcard according to letsencrypt requirements and not in the additional domains field of the certificate. So MC still do not know the fqdn and use the wildcard name with stripped *. as FQDN. This results into wrong links, agents connecting issues, etc.

Proposal to solve this:
The default MC site needs to have the knowledge of the used FQDN. The additional customers FQDN can already being specified in the configuration with DNS item.
If in config reverse proxy is configured and DNS item (this is a new config item) is specified in the default site configuration section. Then this can more easily being used in the code instead of trying to fetch the FQDN from a not getting certificate from reverse proxy.

**Comment by hunter-nl** [comment]
With your recent ‘fix’. How does MC server knows which hostname it should being added to the request to reverse proxy? Or do you use the system hostname.
In my case this is an internal hostname and reverse proxy does not have this knowledge and certainly not in any certificates.

**Comment by Ylianst** [comment]
For the fix I just put in, the domain "dns" setting is used to query the reverse proxy. If there is no domain "dns" in the domain, or it's the default domain then the settings "cert" value is used. That should fix problem #1.

Note that I don't expect you to load the wildcard certificate into MeshCentral, it should only be on the reverse proxy.

**Comment by hunter-nl** [comment]
If the wildcard certificates (I place these manually by script on MC server) is not loaded by MC server then hash errors occurs. I can disable this in the configuration of MC. But that results into less secure communication with agents.

**Comment by Ylianst** [comment]
I need to work on your second problem now. This is a real issue. There is a related issue that if the reverse proxy updates the certificate, MeshCentral will be stuck until you restart it. I will work requesting the certificates hashes after the web server start and look at what to do for the cert update issue.

For your third issue, I am not sure I get this because you should not be loading the reverse-proxy cert into MeshCentral.

For your solution: I do understand that MeshCentral needs to know it's DNS name (that is important). This is what the "cert" setting it for (maybe I should also accept the name "dns" in the setting section as equivalent to make it clear?).

**Comment by Ylianst** [comment]
In answer to this: "If the wildcard certificates (I place these manually by script on MC server) is not loaded by MC server then hash errors occurs. I can disable this in the configuration of MC. But that results into less secure communication with agents."

The alternative is to to this:

```
{
  "settings": {
    "cert": "default.sample.com",
    "tlsOffload": "192.168.0.100"
  },
  "domains": {
    "": {
      "title": "Default"
      "certUrl" : "file://myProxyCert1.cer"
    },
    "customer1": {
      "title": "Customer1",
      "dns" : "customer1.sample.com",
      "certUrl" : "file://myProxyCert1.cer"
    }
  }
}
```

This will get around your second problem until I get a real fix for it.

**Comment by hunter-nl** [comment]
For my third issue: So far I discovered in your code is that you still use CN from the certificates. That is the wildcard fqdn entry. So then you still can not discovery the to be used fqdn. And then the only solution is that you need a DNS manually config item. Then use this to fetch the certificate hashes. But this needs to be done after MC https server is already up and running. It can take up to 1 minute (time is configurable in haproxy) before reverse proxy allow ‘healthy’ backend server.
During typing this response you provide file:// option. I need to test this tomorrow. But I think my third problem will still exist then.

**Comment by Ylianst** [comment]
Ok, I just fixed the server to make sure it uses the "cert" setting name first.

Current version: It was using the certificate CN first unless it started with "*" when it used the "cert" setting.

New version: Use the "cert" setting. If the "cert" setting it not set, use the certificate CN unless it starts with *.

Working on the timing issue now...

**Comment by Ylianst** [comment]
Published MeshCentral v0.4.3-q with a big upgrade for people using reverse-proxies. MeshCentral will try to load the certificate from the reverse proxy as usual with the "CertUrl", but it will not wait for the response and go ahead and setup the main HTTPS server.

On top of this, if connecting agents report seeing a different web certificate then the one expected by the server, MeshCentral will re-poll the reverse proxy for the certificate. The re-poll can only happen as most every 2 minutes. If MeshCentral detects that the reverse-proxy now has a different certificate, it will disconnect all agents that are holding because they have a bad certificate, causing them to reconnect correctly.

In the screen shot below, I forced the server not to make the initial certificate poll and so, it had incorrect values. The first agent connects and there is a mismatch causing the server to poll the reverse proxy, correct the cert, drop any agents that reported the same problem and start receiving all agent connections correctly.

Feedback appreciated on this. It should fix the timing issue and fix the problem where the reverse proxy updates it's cert.

![x](https://user-images.githubusercontent.com/1319013/67817426-91235500-fa6a-11e9-9bda-df8c2268fe45.png)

**Comment by hunter-nl** [comment]
I've tested v0.4.3-q. This looks 👍. I did not test the certificate change. The next certificate update will be about 4 weeks.
OSX agents still has Desktop issues (Not related to this issue). Messenger is failing with some (win/osx/linux) agents (waiting forever for connection). In trace I see only this difference:
```
WEBREQUEST: /messenger?id=meshmessenger/node%2F%2FiPPNuIGF612OY3H%24LlRka%244%406WFQLeusoLhuXZ83gnax4ANymcWs3bE5WBoOF%401Y/user%2F%2F ...
WEBREQUEST: /meshrelay.ashx/.websocket?id=meshmessenger/node//iPPNuIGF612OY3H$LlRka$4@6WFQLeusoLhuXZ83gnax4ANymcWs3bE5WBoOF@1Y/user// ...
```
Through websocket looks fine (so agent is connecting back (but no browser tab will be opened).
The other initiator request is html encoded. Maybe MC can not make the match between these two?
